Feature-auth-service

AmjadMassaoud opened the pull request:
refactor: Refactor auth-service, implement necessary changes

Revised `README.md` to accurately reflect current service features, API endpoint definitions, project structure, and setup instructions.

AmjadMassaoud committed:
- fixed files structure

AmjadMassaoud committed:
- removed unnecessary files

AmjadMassaoud committed:
implement auth service with user registration, login, password reset, and token management

AmjadMassaoud committed:
feat(auth-service): update cookie-parser to version 1.4.7 and refactor authentication logic

- Added cookie-parser version 1.4.7 with updated LICENSE, README, and HISTORY files.
- Refactored authentication provider to use updated cookie settings for refresh tokens.
- Implemented password reset functionality with token generation and validation.
- Removed obsolete routes and types related to password management and email verification.
- Updated token verification logic to align with new user model structure.

AmjadMassaoud committed:
feat: integrate Google authentication with OAuth2 flow

- Added Google authentication provider to handle OAuth2 callback and token exchange.
- Implemented user creation and token generation for new Google users.
- Updated package.json to include google-auth-library dependency.
- Added utils for handling URL operations in the whatwg-url package.
- Updated package-lock.json to reflect new dependencies and versions.
- Enhanced access token verification process with logging for debugging.

AmjadMassaoud committed:
docs: add README for auth service with features, setup, and usage examples

AmjadMassaoud committed:
feat(auth-service): add PasswordResetTokenEntity and update password reset logic

AmjadMassaoud committed:
docs: update README with enhanced features, API endpoints, and usage examples

AmjadMassaoud committed:
feat(auth-service): enhance password reset token logic to return existing token if already issued

AmjadMassaoud committed:
chore(auth-service): update dependencies to include type definitions for express and redis

AmjadMassaoud committed:
feat(auth-service): implement Redis-based password reset token management and limit reset attempts

AmjadMassaoud committed:
Update README.md

AmjadMassaoud committed:
Refactor: Removed unwanted code

AmjadMassaoud committed:
Merge branch 'feature-auth-service' of https://github.com/AmjadMassaoud/Innogram into feature-auth-service

AmjadMassaoud committed:
Update .gitignore

Added node_module file to .gitignore file

AmjadMassaoud committed:
Delete apps/auth-service/node_modules directory

Deleted node_modules from auth-service

AmjadMassaoud committed:
Delete .vscode directory

- Deleted .vscode file

AmjadMassaoud committed:
Delete apps/chats-service directory

AmjadMassaoud committed:
Delete apps/notifications-service directory

AmjadMassaoud committed:
Delete apps/posts-service directory

AmjadMassaoud committed:
Delete apps/users-service directory

AmjadMassaoud committed:
Update auth.controller.ts

Change route name for consistency

AmjadMassaoud committed:
Update password-reset-token.entity.ts

AmjadMassaoud committed:
Update token-entity.ts

AmjadMassaoud committed:
Rename token-entity.ts to user-auth-entity.ts

AmjadMassaoud committed:
Update auth.provider.ts

AmjadMassaoud committed:
Update tsconfig.json

removed unnecessary comments

AmjadMassaoud committed:
Create registration-method.ts

AmjadMassaoud committed:
Update user-auth-entity.ts

AmjadMassaoud committed:
feat(auth-service): refactor authentication logic and enhance error handling

AmjadMassaoud committed:
Merge branch 'feature-auth-service' of https://github.com/AmjadMassaoud/Innogram into feature-auth-service

AmjadMassaoud committed:
feat: Implement OpenAPI documentation for auth service

refactor: Refactor auth-service, implement necessary changes

Revised `README.md` to accurately reflect current service features, API endpoint definitions, project structure, and setup instructions.

AmjadMassaoud committed:
fix: catching errors

Errors were not catched on the controller level.

AmjadMassaoud committed:
refactor: Make cookie available to all paths on the domain

AmjadMassaoud committed:
refactor: remove unecessary line

AmjadMassaoud committed:
refactor: add refreshToken to be sent with response on signup

AmjadMassaoud committed:
feat: exposed Auth-service to API-Gayeway.

AmjadMassaoud commented:
**feat(gateway): Implement Auth Service Gateway Layer**

**About**
This PR introduces the first major piece of our API Gateway's functionality by setting up the entire authentication layer. The main idea is to make our api-gateway the single, official "front door" for all things related to user accounts.

Instead of clients needing to know about our internal auth-service, they'll just talk to the gateway. This allows the gateway to handle the tricky parts of the auth flow, like managing cookies and routing requests, keeping our frontend and other services much simpler and more secure.

**Changes**
Here’s a look at what’s been added to make this happen:

Centralized Communication with AuthHttpProvider: I've created a new provider (auth.http.provider.ts) to handle all HTTP calls to the downstream auth-service. This keeps our communication logic in one place, so our controllers stay clean and we don't repeat ourselves. It also includes a reusable error handler to make sure we're consistent with how we manage issues from the auth-service.

**New Auth & Password Controllers:**

AuthGatewayController handles the main auth routes (/register, /login, /logout, etc.). A key job it has is taking the Set-Cookie header for the refresh token from the auth-service and properly relaying it back to the user's browser.

AuthPasswordGatewayController is dedicated to the password reset flow (/request-password-reset, /reset-password), keeping that logic nicely separated.

Making Routes Public: I've used our @PublicRoute() decorator on endpoints like login and registration so that they correctly bypass the global authentication guard, allowing new users to sign up and existing users to log in.

**Testing**
Get the Services Running:

First, make sure both the api-gateway and the auth-service are running from the root of the monorepo.

npm run start:dev api-gateway

npm run start:dev auth-service (or the equivalent command for your Express service)

Test the Login Flow:

In Postman (or your favorite API client), send a POST request to http://localhost:3002/api-gateway/auth/login with some valid user credentials.

Check the response: You should get back an accessToken in the JSON body.

Check the headers: Look in the "Headers" tab of the response for a Set-Cookie header.

Check Postman's cookie jar: Click the "Cookies" link under the request URL in Postman. You should see the jid refresh token stored there for localhost.

Test the Refresh Flow:

Now, send an empty POST request to http://localhost:3002/api-gateway/auth/refresh-token. Postman will automatically include the jid cookie.

Check the response: You should get back a brand new accessToken.

UladzislauKudravets commented on apps/auth-service/README.md:
```diff
@@ -0,0 +1,107 @@
+- `POST /innogram/password/request-reset` — Request password reset
+- `POST /innogram/password/reset` — Reset password with token
+
+## Environment Variables
```
you can store it within .env.example file (without secrets ofc)

UladzislauKudravets commented on apps/auth-service/app.ts:
```diff
@@ -0,0 +1,40 @@
+app.use(express.json());
+
+app.use('/innogram/auth', authController);
+app.use('/innogram/password', passwordRouter);
```
so we have authController and passwordRouter?) naming should be consistent

UladzislauKudravets commented on apps/auth-service/controllers/password.controller.ts:
```diff
@@ -0,0 +1,175 @@
+ *         description: Internal Server Error - An error occurred during password reset.
+ */
+router.post('/reset', async (req: Request, res: Response) => {
+  const { error, value } = passwordResetSchema.validate(req.body);
```
you're right! it's really good idea to add some validation. As soon as your application grows consider using separated middleware for validation

UladzislauKudravets commented on apps/auth-service/utils/password.util.ts:
```diff
@@ -0,0 +1,54 @@
+import crypto from 'node:crypto';
```
I see that files in utils directory are not really utils. The logic inside looks like services actually - redis.service, password.service. Example of util is something like function that will generate a random number, or extract domain from url. It's something smaller and more particular, that can be reused in many applications.

AmjadMassaoud replied:
Honestly redis.util and token.util appear more the be services so I moved them to a services dir. But password.util appears more to be a utility file!

UladzislauKudravets commented on apps/auth-service/app.ts:
```diff
@@ -0,0 +1,40 @@
+import express from 'express';
```
it's cool that you're using different directories for different tasks. I like this project structure more than previous one, but it surely can be improved) you anyway should have src directory on top level of your microservice. Also it's good to have dirs like common, shared where you can store other dirs like enums, interfaces, schemas, etc.

UladzislauKudravets commented on apps/auth-service/entities/password-reset-token.entity.ts:
```diff
@@ -0,0 +1,26 @@
+@Entity('password-reset-tokens')
+export class PasswordResetTokenEntity {
+  @ObjectIdColumn()
+  id!: string;
```
seems like fields id, createdAt, updatedAt could be moved into some TypeormBaseEntity class

AmjadMassaoud committed:
refactor: rename files and update imports for better clarity and organization

AmjadMassaoud committed:
feat:
- add .env.example file
- add Dockerfile for each service docker-compose at the root of the project

refactor:
- update imports for auth service structure
- update README.md for better readability and understanding

AmjadMassaoud committed:
refactor: rename verify-access-token.guard for improved clarity

AmjadMassaoud committed:
feat: add docker-compose configuration for api-gateway and auth-service

AmjadMassaoud committed:
fix: add module type to package.json for ES module support to fix httpStatus error.
